clients/src/main/java/org/apache/kafka/clients/RequestCompletionHandler.java

```diff
@@ -24,4 +24,11 @@
 
     void onComplete(ClientResponse response);
 
+    /**
+     * Fire when the request transmission hits a fatal exception.
+     *
+     * @param exception the thrown exception
+     */
+    default void onFailure(RuntimeException exception) {
```

Comment:
I am not so sure about adding this to `RequestCompletionHandler`. This interface is mainly serving `NetworkClient`, but there are no changes to `NetworkClient` here.

Reply:
We do have a case in `ConsumerNetworkClient` which adds an `onFailure` callback. To me it makes sense to include it as part of the RequestCompletionHandler interface.

Reply:
Here is the interface for `KafkaClient`:
```
 ClientRequest newClientRequest(String nodeId,
                                   AbstractRequest.Builder<?> requestBuilder,
                                   long createdTimeMs,
                                   boolean expectResponse,
                                   int requestTimeoutMs,
                                   RequestCompletionHandler callback);
```
It is misleading to add an `onFailure` callback to `RequestCompletionHandler` if it is not going to be used by `KafkaClient` implementations such as `NetworkClient`. The usage in `ConsumerNetworkClient` is different because it is internal. In general, we should avoid leaking implementation details up to the interfaces.

Reply:
I see, if this is the case, we need a customized completion handler for both forwarding and AlterISR IMHO.